packages/context/src/context-view.ts

```diff
@@ -0,0 +1,129 @@
+import {resolveList} from './value-promise';
+const debug = debugFactory('loopback:context:view');
+
+/**
+ * `ContextView` provides a view for a given context chain to maintain a live
+ * list of matching bindings and their resolved values within the context
+ * hierarchy.
+ *
+ * This class is the key utility to implement dynamic extensions for extension
+ * points. For example, the RestServer can react to `controller` bindings even
+ * they are added/removed/updated after the application starts.
+ *
+ */
+export class ContextView<T = unknown> implements ContextEventListener {
+  protected _cachedBindings: Readonly<Binding<T>>[] | undefined;
+  protected _cachedValues: T[] | undefined;
+  private _subscription: Subscription | undefined;
+
+  constructor(
+    protected readonly ctx: Context,
+    public readonly filter: BindingFilter,
+  ) {}
+
+  /**
+   * Start listening events from the context
+   */
+  open() {
+    debug('Start listening on changes of context %s', this.ctx.name);
+    return (this._subscription = this.ctx.subscribe(this));
+  }
+
+  /**
+   * Stop listening events from the context
+   */
+  close() {
+    debug('Stop listening on changes of context %s', this.ctx.name);
+    if (this._subscription && !this._subscription.closed) {
+      this._subscription.unsubscribe();
+      this._subscription = undefined;
+    }
+  }
+
+  /**
+   * Get the list of matched bindings. If they are not cached, it tries to find
+   * them from the context.
+   */
+  get bindings(): Readonly<Binding<T>>[] {
+    debug('Reading bindings');
+    if (this._cachedBindings == null) {
+      this._cachedBindings = this.findBindings();
+    }
+    return this._cachedBindings;
+  }
+
+  /**
+   * Find matching bindings and refresh the cache
+   */
+  protected findBindings() {
+    debug('Finding matching bindings');
+    this._cachedBindings = this.ctx.find(this.filter);
+    return this._cachedBindings;
+  }
+
+  /**
+   * Listen on `bind` or `unbind` and invalidate the cache
+   */
+  listen(event: ContextEventType, binding: Readonly<Binding<unknown>>) {
+    this.reset();
+  }
+
+  /**
+   * Reset the view by invalidating its cache
+   */
+  reset() {
+    debug('Invalidating cache');
+    this._cachedBindings = undefined;
+    this._cachedValues = undefined;
+  }
+
+  /**
+   * Resolve values for the matching bindings
+   * @param session Resolution session
+   */
+  resolve(session?: ResolutionSession) {
+    debug('Resolving values');
+    // We don't cache values with this method
+    return resolveList(this.bindings, b => {
+      return b.getValue(this.ctx, ResolutionSession.fork(session));
+    });
+  }
+
+  /**
+   * Get the list of resolved values. If they are not cached, it tries to find
+   * and resolve them.
+   */
+  async values(): Promise<T[]> {
+    debug('Reading values');
+    // Wait for the next tick so that context event notification can be emitted
+    await promisify(process.nextTick)();
```

Comment:
_reposting an older comment: https://github.com/strongloop/loopback-next/pull/2122/files#r245917341_

Please extract the `promisify` part outside of this function, so that `promisify(process.nextTick)` is called only once. Calling promisify for every call of `values()` seems like preliminary pessimisation to me.

```ts
const nextTick = promisify(process.nextTick);
```

Reply:
Fixed